### ISSUE TITLE
keadm join:PullImage from image service failed

### ISSUE THREAD

**Issue by ST-q** [kubeedge/kubeedge#6017]
**What happened and what you expected to happen**:
My edge device is Jetson Orin NX, and I am using nvidia-container-runtime. When I run the keadm join command, I am unable to pull the image, but my network is working fine. How can I resolve this issue? Alternatively, how can I configure it to use local images, or deploy EdgeCore using the binary deployment method? Or should I upload the images to a private registry？
![image](https://github.com/user-attachments/assets/3c34d4e9-63c1-4c93-965f-aa770addd565)

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):1.30.5
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):1.18.1

**Comment by fisherxu** [comment]
https://kubeedge.io/docs/faq/setup#keadm-join-pull-image-failed-in-edge-nodes   This FAQ may help you, ptal :)

**Comment by ST-q** [comment]
> https://kubeedge.io/docs/faq/setup#keadm-join-pull-image-failed-in-edge-nodes This FAQ may help you, ptal :)

this is my repository
![image](https://github.com/user-attachments/assets/25dd4122-12eb-4f7a-af40-4bd69741b857)

My command format has always been incorrect. Can you tell me what the correct command format should be?
![image](https://github.com/user-attachments/assets/55719758-76c4-437e-85ad-168bc03b6720)

I have try:
sudo keadm join --cloudcore-ipport=192.168.0.1:10000 --image-repository=--image-repository=tutowang/installation-package:v1.18.1 --token=***

sudo keadm join --cloudcore-ipport=192.168.0.1:10000 --image-repository=https://hub.docker.com/repository/docker/tutowang  --token=***

sudo keadm join --cloudcore-ipport=192.168.0.1:10000 --image-repository=tutowang/installation-package:v1.18.1 --kubeedge-version=v1.18.1 --token=***

 sudo keadm join --cloudcore-ipport=192.168.0.1:10000 --image-repository=https://hub.docker.com/repository/docker/tutowang --kubeedge-version=v1.18.1 --token=***

**Comment by Shelley-BaoYue** [comment]
@ST-q  Can you pull this image manually successfully?

**Comment by ST-q** [comment]
> @ST-q Can you pull this image manually successfully?
Yes, I can manually pull the image-v1.18.1, but I'm not sure if it's an architectural issue. My device is an Orin-NX, and when I try to use keadm join, it doesn't prioritize using the local image. It still attempts to pull the image, which results in a network error.

**Comment by ST-q** [comment]
> @ST-q Can you pull this image manually successfully?

@Shelley-BaoYue  Yes, I can manually pull the image-v1.18.1, but I'm not sure if it's an architectural issue. My device is an Orin-NX, and when I try to use keadm join, it doesn't prioritize using the local image. It still attempts to pull the image, which results in a network error.

**Comment by ST-q** [comment]
> @ST-q Can you pull this image manually successfully?

@Shelley-BaoYue  Yes, I can manually pull the image-v1.18.1, but I'm not sure if it's an architectural issue. My device is an Orin-NX, and when I try to use keadm join, it doesn't prioritize using the local image. It still attempts to pull the image, which results in a network error.

**Comment by Shelley-BaoYue** [comment]
keadm join will first use local image.  You need pull the image to k8s.io namespace of edge runtime like `ctr -n k8s.io image pull xxx`.

**Comment by ST-q** [comment]
> keadm join will first use local image. You need pull the image to k8s.io namespace of edge runtime like `ctr -n k8s.io image pull xxx`.

@Shelley-BaoYue  Thank you for your reply.In fact, in my edge environment with Ubuntu 20.04 and Docker 24.07, pulling an image does not require specifying a namespace.
![image](https://github.com/user-attachments/assets/6a99e07e-8909-45ba-a268-a583777e0cd0)

In fact, you can specifically see that I have the following images locally.
![image](https://github.com/user-attachments/assets/533e7322-b1e8-4164-bacf-0f50e4c857ca)

It seemed not to work at the time, and I checked the documentation which said that the image should be pulled from Docker Hub, so this left me quite puzzled.
![image](https://github.com/user-attachments/assets/36f3f6a8-8c95-4b58-93bb-d658e1a44030)
![image](https://github.com/user-attachments/assets/a7b1cfd7-7bb4-4e41-8ba8-5dd8b6902af4)

**Comment by Shelley-BaoYue** [comment]
keadm join use contained as default edge runtime. If you need using docker, please ref to: https://kubeedge.io/docs/setup/prerequisites/runtime#docker-engine

**Comment by ST-q** [comment]
> keadm join use contained as default edge runtime. If you need using docker, please ref to: https://kubeedge.io/docs/setup/prerequisites/runtime#docker-engine

@Shelley-BaoYue  Thank you for your reply. I have reinstalled cri-docker using the precompiled binary file and installed CNI using sudo apt-get install -y kubernetes-cni. I also checked the install_cni_plugins function and added the configuration file at /etc/cni/net.d/10-containerd-net.conflist without making any changes to the configuration content. This time, edgecore was able to run successfully, but it seems that the edge node and the master node are unable to communicate. I wonder if there are any settings that I haven't set up correctly. 
CNI:
![image](https://github.com/user-attachments/assets/b64f3325-40ce-4f7f-9ca9-ffeab810f291)
![image](https://github.com/user-attachments/assets/dbcacb5f-e097-45f1-98b4-0956c57a092a)

error:
![image](https://github.com/user-attachments/assets/b33c6ace-ba93-40b8-987d-229a12ed9a5b)

**Comment by Shelley-BaoYue** [comment]
You can follow this command to visit the edgecore logs and find the reason. 
![image](https://github.com/user-attachments/assets/5c4a4462-b852-45d3-92de-ea50094cc9f3)

**Comment by ST-q** [comment]
> You can follow this command to visit the edgecore logs and find the reason. ![image](https://private-user-images.githubusercontent.com/48788150/394291914-5c4a4462-b852-45d3-92de-ea50094cc9f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4Mzc2MTYsIm5iZiI6MTczMzgzNzMxNiwicGF0aCI6Ii80ODc4ODE1MC8zOTQyOTE5MTQtNWM0YTQ0NjItYjg1Mi00NWQzLTkyZGUtZWE1MDA5NGNjOWYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDEzMjgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxOGNmZGRiOTg0ZWU0ZGRkZTQyYjllNGQ4ZDlkYzA5ZGMzOWMyOGI3Y2JkZDBkZTJlMjRmMDRjZmIwN2I1YzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pFxDuNWxyH1bFPqLnCm3F9NUVy5D_2q_m15We52VegA)

@Shelley-BaoYue  I sincerely thank you for your prompt response. I have rechecked the content of the FAQ and reviewed the edge logs and cloud logs. The content of the cloud logs is quite extensive and I find it a bit difficult to understand, but it seems that the services have started up normally. Here is the specific information:
edge logs:
![image](https://github.com/user-attachments/assets/ec4bde2d-c550-46ae-ad06-fed4be805ec3)

cloud logs:
![image](https://github.com/user-attachments/assets/a69671d2-d5cf-449d-82f5-9324260201a2)
![image](https://github.com/user-attachments/assets/1c0e5943-80a9-4243-99ae-a0b3a726ddc2)
![image](https://github.com/user-attachments/assets/4af92668-2cd0-475f-b106-174ffed56a97)

FAQ:
![image](https://github.com/user-attachments/assets/1ea73ea6-4180-476b-b6ea-ceaae9dcd9d0)
Regarding the first point, I used the following command on the cloud side: keadm init --advertise-address=192.168.0.1 --kubeedge-version=v1.18.1 --kube-config=$HOME/.kube/config, so the IP should be within the advertise-address list. For the second point, my edge device connects to my cloud device via SSH and shares the network through a physical network interface, both being on the same local area network (LAN), and the edge device can access this IP. As for the third point, I am able to start cloudcore normally, with the specific information as follows. For the fourth point, I am not sure how to check the content of these logs.

cloudcore success:
![image](https://github.com/user-attachments/assets/5133fd14-9c67-4f35-aaf9-d9183a56ab7f)

**Comment by Shelley-BaoYue** [comment]
Manually curl advertise-address:10002,  check whether this ip:port can be accessed.

**Comment by ST-q** [comment]
> Manually curl advertise-address:10002, check whether this ip:port can be accessed.

@Shelley-BaoYue Thank you very much, your timely response has been extremely helpful to me. I have conducted a port test, and here is the information. I encountered a connection refusal, which I suspect is due to the absence of a service listening on the port. I realized that my cloudcore is running in container mode. Upon re-reading the relevant content in the FAQ, I noticed that when I initialized with keadm init --advertise-address=${THE-EXPOSED-IP}, I did not specify the port, it should have been ${THE-EXPOSED-IP}:10000. Therefore, I rechecked the actual node IP of the master and performed port testing and join testing for the edge node, but there seem to still be issues. Could you help me check the problem? Thank you very much.

curl -v http://192.168.0.1:10002/ ：
![image](https://github.com/user-attachments/assets/5b068fab-2851-4480-bece-698af7e5cf0c)

cloudcore.yaml
![image](https://github.com/user-attachments/assets/7f52325f-ce4b-45d5-9589-1ef4334ce65a)

container ip:
![image](https://github.com/user-attachments/assets/7f6a40de-5bdc-4ecb-862a-22d49070a416)

keadm join logs:
![image](https://github.com/user-attachments/assets/31fce3d8-5ff9-4e34-872d-03c675f77bdd)

curl -v http://192.168.65.3:10002/ 
![image](https://github.com/user-attachments/assets/389fa8e5-9071-4b4d-a09b-445b587ee483)

FAQ:
![image](https://github.com/user-attachments/assets/57c2071e-62aa-49aa-84df-7baf2977d3ec)

Thank you so much!

**Comment by ST-q** [comment]
> Manually curl advertise-address:10002, check whether this ip:port can be accessed.

@Shelley-BaoYue  Thank you for your reply. After reviewing the FAQ and my configuration again, I found that my cloudcore is deployed as a container and I have modified it to be a NodePort type service so that I can access it correctly. I used curl to test, and the results are as follows, but it seems that during the keadm join process, it still points to other ports and cannot point to the port I need. How can I modify the command to join correctly? (192.168.0.1 is master ip)

 curl -v http://192.168.0.1:31680/:
![image](https://github.com/user-attachments/assets/cd29d54a-4dcb-4682-908a-d832ce9531aa)

cloudcore:
![image](https://github.com/user-attachments/assets/8a024108-1227-4dc5-ab48-5e057f9b2feb)

join error:
![image](https://github.com/user-attachments/assets/b455844f-795e-44f2-9ff8-b7861395ab5a)

logs:
![image](https://github.com/user-attachments/assets/75dc4504-e25e-43f3-a198-1df3a76a7730)

**Comment by Shelley-BaoYue** [comment]
You can try using `keadm join -h` to find how to get `keadm join --set` to modiy the port which default value is 10000-10004.

**Comment by ST-q** [comment]
> You can try using `keadm join -h` to find how to get `keadm join --set` to modiy the port which default value is 10000-10004.

@Shelley-BaoYue Hello, thank you very much for your reply. I am encountering a few issues now. I have a local computer with an internal IP of 192.168.0.1, and I am sharing the network with Jetson via an Ethernet cable; the Jetson's internal IP is 192.168.0.6. I have created a Kubernetes cluster on my local computer using Docker Desktop, and the node's IP address is 192.168.65.3. It seems that the node's IP is not in the same subnet as my actual IP, and the cluster's IP is on a network isolated from the local host. So, I should provide the --advertise-address parameter as 192.168.0.1, and open ports 10000-10002, and configure NAT to enable communication between the cluster and my local network. Is that correct? (host: wsl2-ubuntu20.04 ; jetson-orin-nx: ubuntu20.04)
